server/api/util.go

```diff
@@ -23,20 +23,44 @@ import (
 	"github.com/juju/errors"
 	"github.com/pingcap/pd/pkg/apiutil"
 	"github.com/pingcap/pd/server"
+	"github.com/pingcap/pd/server/error_code"
+	log "github.com/sirupsen/logrus"
 	"github.com/unrolled/render"
 )
 
-func readJSONRespondError(r *render.Render, w http.ResponseWriter, body io.ReadCloser, data interface{}) error {
+// Respond to the client about the given error.
+// By default an error is assumed to be a 400 Bad Request
+// unless it is recognized as an ErrorCode
+// If the error is nil, this responds with a 500
+func errorResp(rd *render.Render, w http.ResponseWriter, err error) {
+	if err == nil {
+		log.Errorf("nil given to errorResp")
+		rd.JSON(w, http.StatusInternalServerError, "nil error")
+		return
+	}
+	if errCode, ok := errors.Cause(err).(errcode.ErrorCode); ok {
+		w.Header().Set("TiDB-Error-Code", string(errCode.Code()))
```

Comment:
Looks like we should export `errorResp()` and update all API to use this method to return error to client. Then we can guarantee that all error returns will contain error codes in the custom header.

Reply:
Okay, switched to 500. So I think we can use this everywhere now. I wanted to limit the work of this PR to the minimal that is needed to demonstrate the soundness of this approach. This avoids merge conflicts and otherwise makes it easy for us to make changes to this PR. We can add follow-up PRs to switch over more errors to specify codes.